Blead breaks build in g++, clang++ builds on some platforms

Compiling `perl` using `g++` or `clang++` has started to fail on certain platforms, such as several of the BSDs.

For example, building on FreeBSD-13 with the following configuration:
```
$ g++ --version | head -n1
g++ (FreeBSD Ports Collection) 12.2.0

$ sh ./Configure -des -Dusedevel -Dcc=g++ && make test_prep
```
... the build stops here:
```
g++ -c -DPERL_CORE -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -DNO_POSIX_2008_LOCALE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_FORTIFY_SOURCE=2 -O2 -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wwrite-strings -Wno-use-after-free locale.c
locale.c: In function 'void S_ints_to_tm(tm*, int, int, int, int, int, int, int, int, int)':
locale.c:8072:25: error: invalid conversion from 'const char*' to 'char*' [-fpermissive]
 8072 |         mytm->tm_zone = zone;
      |                         ^~~~
      |                         |
      |                         const char*
*** Error code 1

Stop.
make: stopped in /usr/home/jkeenan/gitwork/perl
```
Line 8072 in `locale.c` was last modified in blead yesterday.
```
commit 36e41d53ff6a5995c16234f8027ef34b5e6a34cc
Author:     Karl Williamson <khw@cpan.org>
AuthorDate: Wed Jan 10 11:30:11 2024 -0700
Commit:     Karl Williamson <khw@cpan.org>
CommitDate: Wed Feb 7 10:59:05 2024 -0700

    Don't use both mini_mktime() and mkttime()
```
`g++` builds on Linux so far appear to be unaffected.

@khwilliamson, can you take a look?

On the same FreeBSD-13 machine as above, we get a similar build failure using `clang++` instead of `g++`:
```
$ clang++ --version
FreeBSD clang version 14.0.5 (https://github.com/llvm/llvm-project.git llvmorg-14.0.5-0-gc12386ae247c)
Target: x86_64-unknown-freebsd13.2
Thread model: posix
InstalledDir: /usr/bin

clang++ -c -DPERL_CORE -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -DNO_POSIX_2008_LOCALE -I/usr/local/include -O2 -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-declaration-after-statement -Wc++-compat -Wwrite-strings locale.c
clang++: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]
locale.c:8072:25: error: assigning to 'char *' from 'const char *' discards qualifiers
        mytm->tm_zone = zone;
                        ^~~~
1 error generated.
*** Error code 1

Stop.
```

--- a/locale.c
+++ b/locale.c
@@ -8044,11 +8044,8 @@ S_ints_to_tm(pTHX_ struct tm * mytm,
 
         /* But since mini_mktime() doesn't handle these fields, save the
          * results we already have for them */
-#  ifdef HAS_TM_TM_GMTOFF
-        long int gmtoff = mytm->tm_gmtoff;
-#  endif
-#  ifdef HAS_TM_TM_ZONE
-        const char *zone = mytm->tm_zone;
+#  if defined(HAS_TM_TM_GMTOFF) || defined(HAS_TM_TM_ZONE)
+        struct tm mytm2 = *mytm;
 #  endif
 
         /* Repeat the entire process with mini_mktime */
@@ -8066,10 +8063,10 @@ S_ints_to_tm(pTHX_ struct tm * mytm,
 
         /* And use the saved libc values for tm_gmtoff and tm_zone */
 #  ifdef HAS_TM_TM_GMTOFF
-        mytm->tm_gmtoff = gmtoff;
+        mytm->tm_gmtoff = mytm2.tm_gmtoff;
 #  endif
 #  ifdef HAS_TM_TM_ZONE
-        mytm->tm_zone = zone;
+        mytm->tm_zone = mytm2.tm_zone;
 #  endif
 
     }

